### ISSUE TITLE
Is it possible to build up an connection between two ROS2 nodes via internet?

### ISSUE THREAD

**Issue by Alex031544** [ros2/ros2#328]
Hello,

as written in the title - for an wider distributed network it is interesting to us to know if it is generally possible to build up a connection via the internet.

Kind regards,
Alex

**Comment by wjwwood** [comment]
**tl;dr** It's possible, but not likely easy to do. Also we do not have a tutorial or anything about how to do it.

---

Since we're using RTPS (DDS wire protocol) the communication is broken into two parts: discovery and then peer-to-peer communication.

The peer-to-peer communication is very much like other point to point services, you simply need to ensure the port being used can traverse any firewalls or NATs. To do this, you need to be able to predict the ports that will be used to make direction connections between publishers and subscribers. This was discussed on discourse:

https://discourse.ros.org/t/ros2-behind-a-firewall/1178/3

The discovery portion is a bit more tricky, since by default it uses multicast UDP. But the spec allows for other mechanisms for discovery, like having a single process that acts as a discovery service (this is like ROS 1's master), which might be a better solution if you needed to do discovery over the internet. I'm not too familiar with different forms of discovery, since we're focused on local area networks for the most part.

---

For both ROS 1 and I would imagine ROS 2 as well, most people will choose to use VPN to funnel data across the internet rather than try to open ports and configure the discovery system. The VPN has the advantage of generally not needing any special configurations for ROS and it encrypts data sent over the VPN, adding some security in the process.

---

As a point of process, you might want to ask questions like this on https://discourse.ros.org/c/ng-ros, and leave this issue tracker for bugs or enhancements. Asking a question here isn't wrong, but I think discourse is more appropriate.

To that end, I'm going to close this issue. Feel free to continue discussion here or to move it to discourse. I can reopen it in the future if it becomes necessary.
